### ISSUE TITLE
[Nix reliability]cache.nixos.org is unreliable

### ISSUE THREAD

**Issue by 0xABAB** [NixOS/nix#1448]
We can see `curl: (22) The requested URL returned error: 500` on Nixpkgs/Nix project managed infrastructure (Travis).

The URL could be downloaded a few minutes after. 

This means that the service cache.nixos.org cannot be assumed to be reliable and Nix must be designed to cope with failure, which AFAIK, it is not. Failing the entire process, just because one file was not present during one request is not a sensible semantics.

```
*** Downloading ‘https://cache.nixos.org/nar/1x0d8f6ma90wfvbrqfjgdmmpjmh4pwscswdp0c9jvir8s2mh82cg.nar.xz’ to ‘/nix/store/bvw5vsgbk374rxdxbasf70b8rpa28qys-libXcursor-1.1.14’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:04 --:--:--     0
curl: (22) The requested URL returned error: 500 
/nix/store/s2hmf51bprsxb20v3bfigs8a7zsw3hjl-xz-5.2.2-bin/bin/xz: (stdin): File format not recognized
error: unexpected end-of-file
download of ‘https://cache.nixos.org/nar/1w7v1smycmljsl0z5052wavp39lclnrcjmmb00qbnb468syp9x9p.nar.xz’ failed: No such file or directory
could not download ‘/nix/store/kajgj85vaqxh770ifmb6v088w9qcg698-libdrm-2.4.81-dev’ from any binary cache
fetching path ‘/nix/store/kajgj85vaqxh770ifmb6v088w9qcg698-libdrm-2.4.81-dev’ failed with exit code 1
fetching path ‘/nix/store/cnqr1j3ss321d5a7b9h3jn5dgjwfdvr9-libXinerama-1.1.3’...
cannot build derivation ‘/nix/store/pga94c9jyl50lll15k0i37z89kxqzbzy-qemu-xen-x86-only-2.9.0.drv’: 1 dependencies couldn't be built
killing process 5904
killing process 5909
killing process 5895
killing process 5894
error: build of ‘/nix/store/pga94c9jyl50lll15k0i37z89kxqzbzy-qemu-xen-x86-only-2.9.0.drv’ failed
The invocation of "nix-build -A qemu_xen -A xen-slim -A gnome3.gnome-boxes -A linuxPackages_4_4.dpdk -A cide -A docker-gc -A libguestfs -A virtinst -A xenPackages.xen_4_8-vanilla -A linuxPackages_4_11.dpdk -A linuxPackages.dpdk -A virtmanager -A linuxPackages.odp-dpdk -A xen -A docker-machine-kvm -A linuxPackages_hardened.odp-dpdk -A linuxPackages_chromiumos_3_18.dpdk -A libvirt-glib -A virtviewer -A linuxPackages.ofp -A linuxPackages_latest_xen_dom0.ofp -A qemu_xen_4_8 -A virt-top -A linuxPackages_hardened.pktgen -A bedup -A libvirt -A btrbk -A linuxPackages_chromiumos_latest.odp-dpdk -A linuxPackages_4_4.ofp -A linuxPackages_4_4.odp-dpdk -A python36Packages.libvirt -A pipework -A perlPackages.SysVirt -A linuxPackages_4_12.ofp -A docker-edge -A docker -A linuxPackages_4_11.pktgen -A btrfs-progs -A linuxPackages_latest_xen_dom0.pktgen -A linuxPackages_4_12.odp-dpdk -A linuxPackages_4_4.pktgen -A linuxPackages_latest_xen_dom0.dpdk -A linuxPackages_hardened.dpdk -A linuxPackages_chromiumos_latest.ofp -A linuxPackages_hardened.ofp -A linuxPackages_4_12.pktgen -A virt-viewer -A linuxPackages_chromiumos_3_18.ofp -A glusterfs -A linuxPackages_chromiumos_latest.dpdk -A collectd -A skopeo -A xenPackages.xen_4_8-slim -A linuxPackages_4_12.dpdk -A ocamlPackages.ocaml_libvirt -A minikube -A linuxPackages_4_11.odp-dpdk -A linuxPackages_latest_xen_dom0.odp-dpdk -A virtmanager-qt -A linuxPackages_chromiumos_3_18.odp-dpdk -A linuxPackages_chromiumos_latest.pktgen -A linuxPackages.pktgen -A python27Packages.libvirt -A btrfsProgs -A panamax_api -A linuxPackages_chromiumos_3_18.pktgen -A linuxPackages_4_11.ofp /home/travis/.nox/nixpkgs" failed
````

**Comment by vcunat** [comment]
There are various options, e.g. `--keep-going`, `--fallback`, depending on what you want...

**Comment by copumpkin** [comment]
Nix 1.12 has auto-retry built into its downloader for a bunch of retriable situations. So basically most HTTP codes in the 5xx range and a few others. You can see the logic for that [here](https://github.com/NixOS/nix/blob/c0015e87af70f539f24d2aa2bc224a9d8b84276b/src/libstore/download.cc#L275-L294).

Note that this will not help downloaders defined inside nixpkgs, but anything that Nix itself downloads (channel bundles, binary cache substitutes, etc.) will magically start retrying sensibly with 1.12.

**Comment by 0xABAB** [comment]
Relevant: https://github.com/NixOS/nix/issues/807

**Comment by domenkozar** [comment]
Closing this as @copumpkin mentions this will be fixed in 1.12
